pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,239 @@
+	"k8s.io/klog/v2"
+)
+
+const (
+	reasonIncompatibleOperatorsInstalled = "IncompatibleOperatorsInstalled"
+	MaxOpenShiftVersionProperty          = "olm.maxOpenShiftVersion"
+	OwnerKindKey                         = "olm.operatorframework.io/owner-kind"
+	OwnerNameKey                         = "olm.operatorframework.io/owner-name"
+	PackageNameKey                       = "olm.operatorframework.io/package-name"
+	BundleNameKey                        = "olm.operatorframework.io/bundle-name"
+	BundleVersionKey                     = "olm.operatorframework.io/bundle-version"
+)
+
+type dynamicUpgradeableConditionController struct {
+	name                   string
+	kubeclient             kubernetes.Interface
+	clusterExtensionClient *clients.ClusterExtensionClient
+	operatorClient         *clients.OperatorClient
+	prefixes               []string
+	logger                 logr.Logger
+}
+
+func NewDynamicUpgradeableConditionController(kubeclient kubernetes.Interface, clusterExtensionClient *clients.ClusterExtensionClient, name string, operatorClient *clients.OperatorClient, eventRecorder events.Recorder, prefixes []string) factory.Controller {
+	c := &dynamicUpgradeableConditionController{
+		name:                   name,
+		clusterExtensionClient: clusterExtensionClient,
+		kubeclient:             kubeclient,
+		operatorClient:         operatorClient,
+		prefixes:               prefixes,
+		logger:                 klog.FromContext(context.TODO()).WithName(name),
+	}
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).WithInformers(operatorClient.Informer(), clusterExtensionClient.Informer().Informer()).ToController(name, eventRecorder)
+}
+
+func (c *dynamicUpgradeableConditionController) sync(ctx context.Context, _ factory.SyncContext) error {
+	c.logger.V(4).Info("sync started")
+	defer c.logger.V(4).Info("done syncing")
+
+	updateStatusFuncs, err := c.ensureControllersManaged()
+	if err != nil {
+		c.logger.V(4).Error(err, "error verifying OLM state")
+		return err
+	}
+
+	incompatibleOperatorsUpdateStatusFuncs, err := c.checkIncompatibleOperators()
+	if err != nil {
+		c.logger.V(4).Error(err, "error checking incompatible operators")
+		return err
+	}
+
+	updateStatusFuncs = append(updateStatusFuncs, incompatibleOperatorsUpdateStatusFuncs...)
+	if _, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, updateStatusFuncs...); updateErr != nil {
+		c.logger.V(4).Info(fmt.Sprintf("Error updating operator condition status: %v", updateErr))
+		return updateErr
+	}
+	return nil
+}
+
+func (c *dynamicUpgradeableConditionController) ensureControllersManaged() ([]v1helpers.UpdateStatusFunc, error) {
+	c.logger.V(4).Info("verifying state of olm controllers")
+	defer c.logger.V(4).Info("done verifying state of olm controllers")
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return nil, err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil, nil
+	}
```

Comment:
Just ran across this code in library-go that seems more robust: https://github.com/openshift/library-go/blob/3c41fd1dea0be2102a6419eb7772bec7ea1b4b38/pkg/operator/staticresourcecontroller/static_resource_controller.go#L287-L289